### PR TITLE
Add request/response telemetry for api client

### PR DESCRIPTION
**Pull request by nronas**
[Ticket](https://deliveroo.atlassian.net/browse/SUPP-2772)

### CODE REVIEW

File: lib/routemaster/middleware/metrics.rb [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+        increment_req_count(request_tags(request_env))
+
+        record_latency(request_tags(request_env)) do
+          @app.call(request_env).on_complete do |response_env|
+            increment_response_count(response_tags(response_env))
+          end
+        end
+      end
+
+      def increment_req_count(tags)
```
Comment (ajaymallya) [inline lib/routemaster/middleware/metrics.rb]: Did you intend these three methods to be part of the public API of the `Metrics` class? Seems like they should be private and `#call` should be the only public method.

**Review by ajaymallya — COMMENTED** [review]
LGTM with one question.

File: lib/routemaster/middleware/metrics.rb [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+      def record_latency(tags, &block)
+        client.time("#{INTERACTION_KEY}.latency", tags: tags) do
+          yield
+        end
+      end
+
+      private
+
+      def can_log?(env)
```
Comment (angetheo) [inline lib/routemaster/middleware/metrics.rb]: `env` in unused?

File: lib/routemaster/middleware/metrics.rb (hunk 1 above)
Comment (nronas) [reply lib/routemaster/middleware/metrics.rb]: 👍

File: spec/routemaster/integration/api_client_spec.rb [hunk 3]
```diff
@@ -128,4 +128,45 @@ class ProcHandler
...
+    let(:metrics_client) { double('MetricsClient') }
+    let(:source_peer) { 'test_service' }
+    let(:url) { "#{host}/success" }
+
+    subject do
+      Routemaster::APIClient.new(metrics_client: metrics_client,
+                                 source_peer: source_peer)
+    end
+
+    context 'when metrics source peer is absentt' do
```
Comment (angetheo) [inline spec/routemaster/integration/api_client_spec.rb]: `absent` 💅

**Review by angetheo — APPROVED** [review]
looks good!

File: lib/routemaster/middleware/metrics.rb [hunk 4]
```diff
@@ -0,0 +1,66 @@
+module Routemaster
+  module Middleware
+    class Metrics
+      INTERACTION_KEY = 'api_client'.freeze
+
+      attr_reader :client, :source_peer
```
Comment (cih) [inline lib/routemaster/middleware/metrics.rb]: Could these also be private?

**Review by cih — APPROVED** [review]

**Pull request merged**